File: packages/guilded/src/incoming.ts
import type { Client } from '@jersey/guildapi';
import type {
	ChatMessage,
	ServerMember,
	Webhook,
} from '@jersey/guilded-api-types';
import { type attachment, cacher, type message } from '@lightning/lightning';

const member_cache = new cacher<`${string}/${string}`, ServerMember>();
const webhook_cache = new cacher<`${string}/${string}`, Webhook>();
const asset_cache = new cacher<string, attachment>(86400000);

export async function fetch_author(msg: ChatMessage, client: Client) {
	try {
		if (!msg.createdByWebhookId) {
			const key = `${msg.serverId}/${msg.createdBy}` as const;
			const author = member_cache.get(key) ?? member_cache.set(
				key,
				(await client.request(
					'get',
					`/servers/${msg.serverId}/members/${msg.createdBy}`,
					undefined,
				) as { member: ServerMember }).member,
			);

			return {
				username: author.nickname ?? author.user.name,
				rawname: author.user.name,
				id: msg.createdBy,
				profile: author.user.avatar,
			};
		} else {
			const key = `${msg.serverId}/${msg.createdByWebhookId}` as const;
			const webhook = webhook_cache.get(key) ?? webhook_cache.set(
				key,
				(await client.request(
					'get',
					`/servers/${msg.serverId}/webhooks/${msg.createdByWebhookId}`,
					undefined,
				)).webhook,
			);

			return {
				username: webhook.name,
				rawname: webhook.name,
				id: webhook.id,
				profile: webhook.avatar,
			};
		}
	} catch {
		return {
			username: 'Guilded User',
			rawname: 'GuildedUser',
			id: msg.createdByWebhookId ?? msg.createdBy,
		};
	}
}

async function fetch_attachments(markdown: string[], client: Client) {
	const urls = markdown.map(
		(url) => (url.split('(').pop())?.split(')')[0],
	).filter((i) => i !== undefined);

	const attachments: attachment[] = [];

	for (const url of urls) {
		const cached = asset_cache.get(url);

		if (cached) {
			attachments.push(cached);
		} else {
			try {
				const signed = (await client.request('post', '/url-signatures', {
					urls: [url],
				})).urlSignatures[0];

				if (signed.retryAfter || !signed.signature) continue;

				attachments.push(asset_cache.set(signed.url, {
					name: signed.signature.split('/').pop()?.split('?')[0] ?? 'unknown',
					file: signed.signature,
					size: parseInt(
						(await fetch(signed.signature, {
							method: 'HEAD',
						})).headers.get('Content-Length') ?? '0',
					) / 1048576,
				}));
			} catch {
				continue;
			}
		}
	}

	return attachments;
}

export async function get_incoming(
	msg: ChatMessage,
	client: Client,
): Promise<message | undefined> {
	if (!msg.serverId) return;

	let content = msg.content?.replaceAll('\n```\n```\n', '\n');

	const urls = content?.match(
		/!\[.*\]\(https:\/\/cdn\.gldcdn\.com\/ContentMediaGenericFiles\/.*\)/gm,
	) ?? [];

	content = content?.replaceAll(
		/!\[.*\]\(https:\/\/cdn\.gldcdn\.com\/ContentMediaGenericFiles\/.*\)/gm,
		'',
	)?.replaceAll(/<(:\w+:)\d+>/g, (_, emoji) => emoji);

	return {
		attachments: await fetch_attachments(urls, client),
		author: {
			...await fetch_author(msg, client),
			color: '#F5C400',
		},
		channel_id: msg.channelId,
		content,
		embeds: msg.embeds?.map((embed) => ({
			...embed,
			author: embed.author
				? {
					...embed.author,
					name: embed.author.name ?? '',
				}
				: undefined,
			image: embed.image
				? {
					...embed.image,
					url: embed.image.url ?? '',
				}
				: undefined,
			thumbnail: embed.thumbnail
				? {
					...embed.thumbnail,
					url: embed.thumbnail.url ?? '',
				}
				: undefined,
			timestamp: embed.timestamp ? Number(embed.timestamp) : undefined,
		})),
		message_id: msg.id,
		plugin: 'bolt-guilded',
		reply_id: msg.replyMessageIds,
		timestamp: Temporal.Instant.from(
			msg.createdAt,
		),
	};
}

File: pkg/platforms/guilded/incoming.go
package guilded

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"path"
	"regexp"
	"strconv"
	"strings"
	"time"

	"codeberg.org/jersey/lightning/pkg/lightning"
)

const assetCacheTTL = 24 * time.Hour

var attachmentRegex = regexp.MustCompile(
	`!\[.*?\]\(https:\/\/cdn\.gldcdn\.com\/ContentMedia(?:GenericFiles)?\/.*\)`,
)

func guildedToLightning(plugin *guildedPlugin, msg *guildedChatMessage) *lightning.Message {
	if msg.ServerID == nil {
		return nil
	}

	if found, _ := plugin.webhookIDsCache.Get(msg.CreatedByWebhookID); found {
		return nil
	}

	attachmentMarkdown := attachmentRegex.FindAllString(msg.Content, -1)

	return &lightning.Message{
		BaseMessage: lightning.BaseMessage{
			EventID:   msg.ID,
			ChannelID: msg.ChannelID,
			Time:      msg.CreatedAt,
		},
		Author:      guildedToLightningAuthor(plugin, msg),
		Content:     attachmentRegex.ReplaceAllString(msg.Content, ""),
		Embeds:      msg.Embeds,
		Attachments: guildedToLightningAttachments(plugin, attachmentMarkdown),
		RepliedTo:   msg.ReplyMessageIDs,
	}
}

func guildedToLightningAuthor(plugin *guildedPlugin, msg *guildedChatMessage) *lightning.MessageAuthor {
	var cacheKey string

	var endpoint string

	if msg.CreatedByWebhookID == "" {
		cacheKey = *msg.ServerID + "/" + msg.CreatedBy
		endpoint = "/servers/" + *msg.ServerID + "/members/" + msg.CreatedBy
	} else {
		cacheKey = *msg.ServerID + "/" + msg.CreatedByWebhookID
		endpoint = "/servers/" + *msg.ServerID + "/webhooks/" + msg.CreatedByWebhookID
	}

	if cachedMember, ok := plugin.membersCache.Get(cacheKey); ok {
		return toAuthor(cachedMember.User, cachedMember.Nickname)
	}

	var responseBody struct {
		Member  *guildedServerMember `json:"member,omitempty"`
		Webhook *guildedUser         `json:"webhook,omitempty"`
	}

	if err := doJSONRequest(plugin.token, http.MethodGet, endpoint, nil, &responseBody); err != nil {
		return &lightning.MessageAuthor{
			Nickname: "Guilded User",
			Username: "GuildedUser",
			ID:       msg.CreatedBy,
			Color:    "#f8d64c",
		}
	}

	if responseBody.Member != nil {
		plugin.membersCache.Set(cacheKey, *responseBody.Member)

		return toAuthor(responseBody.Member.User, responseBody.Member.Nickname)
	}

	plugin.membersCache.Set(cacheKey, guildedServerMember{User: *responseBody.Webhook})

	return toAuthor(*responseBody.Webhook, nil)
}

func toAuthor(user guildedUser, nickname *string) *lightning.MessageAuthor {
	displayName := user.Name
	if nickname != nil {
		displayName = *nickname
	}

	return &lightning.MessageAuthor{
		Nickname:       displayName,
		Username:       user.Name,
		ID:             user.ID,
		ProfilePicture: user.Avatar,
		Color:          "#f8d64c",
	}
}

func guildedToLightningAttachments(plugin *guildedPlugin, markdownLinks []string) []lightning.Attachment {
	attachments := make([]lightning.Attachment, 0, len(markdownLinks))

	for _, mdLink := range markdownLinks {
		url := extractURLFromMarkdown(mdLink)
		if url == "" {
			continue
		}

		if cachedAttachment, ok := plugin.assetsCache.Get(url); ok {
			attachments = append(attachments, cachedAttachment)

			continue
		}

		sig := getSignature(url, plugin.token)
		if sig == nil || len(sig.URLSignatures) == 0 || sig.URLSignatures[0].Signature == nil {
			continue
		}

		urlSignature := *sig.URLSignatures[0].Signature
		attachment := lightning.Attachment{
			Name: path.Base(strings.SplitN(urlSignature, "?", 2)[0]),
			URL:  urlSignature,
			Size: getContentLength(urlSignature),
		}

		plugin.assetsCache.Set(url, attachment)
		attachments = append(attachments, attachment)
	}

	return attachments
}

func extractURLFromMarkdown(markdown string) string {
	start, end := strings.LastIndex(markdown, "("), strings.LastIndex(markdown, ")")
	if start >= 0 && end > start {
		return markdown[start+1 : end]
	}

	return ""
}

func getContentLength(url string) int64 {
	req, err := http.NewRequest(http.MethodHead, url, nil)
	if err != nil {
		return 0
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return 0
	}
	defer resp.Body.Close()

	contentLength, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 64)

	return contentLength
}

func getSignature(url, token string) *guildedURLSignatureResponse {
	var signatureResponse guildedURLSignatureResponse
	if err := doJSONRequest(token, http.MethodPost, "/url-signatures",
		map[string][]string{"urls": {url}}, &signatureResponse); err != nil {
		return nil
	}

	return &signatureResponse
}

func doJSONRequest(token, method, endpoint string, body, out any) error {
	var buf io.Reader

	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("failed to marshal: %w", err)
		}

		buf = bytes.NewReader(b)
	}

	resp, err := guildedMakeRequest(token, method, endpoint, buf)
	if err != nil {
		return err
	}

	defer resp.Body.Close()

	if err = json.NewDecoder(resp.Body).Decode(out); err != nil {
		return fmt.Errorf("failed to decode: %w", err)
	}

	return nil
}
